pyanaconda/modules/payloads/payload/dnf/dnf.py

```diff
@@ -91,11 +102,123 @@ def process_kickstart(self, data):
         source.process_kickstart(data)
         self.add_source(source)
 
+    def _process_kickstart_packages(self, data):
+        """Process the kickstart packages."""
+        packages = PackagesConfigurationData()
+        packages.core_group_enabled = not data.packages.nocore
+        packages.default_environment_enabled = data.packages.default
+
+        if data.packages.environment is not None:
+            packages.environment = data.packages.environment
+
+        packages.packages = list(data.packages.packageList)
+        packages.excluded_packages = data.packages.excludedList
+
+        for group in data.packages.groupList:
+            packages.groups.append(group.name)
+
+            if group.include == GROUP_ALL:
+                packages.groups_package_types[group.name] = GROUP_PACKAGE_TYPES_ALL
+
+            if group.include == GROUP_REQUIRED:
+                packages.groups_package_types[group.name] = GROUP_PACKAGE_TYPES_REQUIRED
+
+        for group in data.packages.excludedGroupList:
+            packages.excluded_groups.append(group.name)
+
+        packages.docs_excluded = data.packages.excludeDocs
+        packages.weakdeps_excluded = data.packages.excludeWeakdeps
+
+        if data.packages.handleMissing == KS_MISSING_IGNORE:
+            packages.missing_ignored = True
+
+        if data.packages.handleBroken == KS_BROKEN_IGNORE:
+            packages.broken_ignored = True
+
+        if data.packages.instLangs == "":
+            packages.languages = RPM_LANGUAGES_NONE
+        elif data.packages.instLangs is not None:
+            packages.languages = data.packages.instLangs
+
+        if data.packages.multiLib:
+            packages.multilib_policy = MULTILIB_POLICY_ALL
+
+        if data.packages.timeout is not None:
+            packages.timeout = data.packages.timeout
+
+        if data.packages.retries is not None:
+            packages.retries = data.packages.retries
+
+        self.set_packages(packages)
+
     def setup_kickstart(self, data):
         """Setup the kickstart data."""
+        self._set_up_kickstart_sources(data)
+        self._set_up_kickstart_packages(data)
+
+    def _set_up_kickstart_sources(self, data):
+        """Set up the kickstart sources."""
         for source in self.sources:
             source.setup_kickstart(data)
 
+    def _set_up_kickstart_packages(self, data):
+        """Set up the kickstart packages."""
+        # The empty packages section won't be printed without seen set to True.
```

Comment:
I'm a bit confused here - do we usually print the `%packages` section if it is empty ? Also, in case setting seen is needed always even for non-empty `%packages` that would seem like a Pykickstart bug.

Reply:
There is a logical difference between an empty `%packages` section and a missing `%packages` section. Pykickstart needs to keep the missing section missing during pre-processing of a kickstart file (for example, flattening), so it uses the `seen` attribute to decide whether to generate the section or not. It is ugly, but this is how it works.